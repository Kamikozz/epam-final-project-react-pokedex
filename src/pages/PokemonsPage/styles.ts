import { createStyles } from "@material-ui/core";

const styles = createStyles({
  root: {
    margin: "20px",
    flexGrow: 1
  },
  actions: {
    justifyContent: "center"
  },
  nextPageButtonWrapper: {
    position: "relative"
  },
  nextPageButton: {
    margin: "10px 0 10px 0",
    backgroundColor: "red",
    "&:hover": {
      backgroundColor: "red"
    }
  },
  nextPageButtonLoader: {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)"
  }
});

export default styles;
