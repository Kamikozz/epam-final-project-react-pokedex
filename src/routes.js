const routes = {
  pokemonsList: "/",
  pokemon: "/pokemons/:id",
  caughtPokemonsList: "/caught_pokemons/"
};

export default routes;
