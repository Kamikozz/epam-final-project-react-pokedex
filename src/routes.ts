const routes = {
  pokemonsPage: "/",
  pokemonPage: "/pokemons/:id",
  caughtPokemonsPage: "/caught_pokemons/"
};

export default routes;
