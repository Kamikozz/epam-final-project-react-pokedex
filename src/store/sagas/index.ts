import { takeEvery, takeLatest } from 'redux-saga/effects';

import { fetchPokemons } from './fetchPokemons';
import { fetchCaughtPokemons } from './fetchCaughtPokemons';
import { fetchNextPagePokemons } from './fetchNextPagePokemons';
import { postAndAddCaughtPokemon } from './postAndAddCaughtPokemon';
import {
  CaughtPokemonsActionType, PokemonsActionType, PaginatedPokemonsActionType,
} from '../slices';

/*
  Starts fetchUser on each dispatched `USER_FETCH_REQUESTED` action.
  Allows concurrent fetches of user.
*/
function* rootSaga() {
  yield takeEvery(PokemonsActionType.FETCH_REQUESTED, fetchPokemons);
  yield takeEvery(CaughtPokemonsActionType.FETCH_REQUESTED, fetchCaughtPokemons);
  yield takeEvery(
    PaginatedPokemonsActionType.FETCH_NEXT_PAGE_POKEMONS_REQUESTED, fetchNextPagePokemons,
  );
  yield takeEvery(CaughtPokemonsActionType.POST_AND_ADD_ONE_REQUESTED, postAndAddCaughtPokemon);
}

// /*
//   Alternatively you may use takeLatest.

//   Does not allow concurrent fetches of user. If "USER_FETCH_REQUESTED" gets
//   dispatched while a fetch is already pending, that pending fetch is cancelled
//   and only the latest one will be run.
// */
// function* rootSaga() {
//   yield takeLatest(ActionType.USER_FETCH_REQUESTED, fetchUser);
// }

export default rootSaga;
