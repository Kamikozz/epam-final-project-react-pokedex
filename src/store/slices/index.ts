export * from './pokemons';
export * from './caughtPokemons';
export * from './users';
export * from './pages';
