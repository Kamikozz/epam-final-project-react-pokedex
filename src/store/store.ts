import { createStore, compose, applyMiddleware } from 'redux';
import { devToolsEnhancer } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import rootSaga from '../store/sagas';

import rootReducer from "../reducer";

// create the saga middleware
const sagaMiddleware = createSagaMiddleware();
const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(sagaMiddleware),
    devToolsEnhancer({}),
  ),
);

// then run the saga
sagaMiddleware.run(rootSaga);

export default store;
