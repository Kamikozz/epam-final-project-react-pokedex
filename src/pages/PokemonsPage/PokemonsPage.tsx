import React, { useEffect, useRef } from "react";
import { useSelector, useDispatch } from 'react-redux';

import { WithStyles, Grid, CardActions, Button } from "@material-ui/core";
import { withStyles } from "@material-ui/core/styles";

import {
  fetchPaginatedPokemons,
  fetchCaughtPokemons,
  fetchNextPagePokemons,
  postAndAddCaughtPokemon,
  selectCaughtPokemonsItems,
  selectPage,
  selectPokemonsItems,
  selectPokemonsIsLoading,
  selectCaughtPokemonsIsLoading,
} from '../../store/slices';

import {
  PokemonItem,
  Loader,
} from "../../components";

import styles from "./styles";

interface Props extends WithStyles<typeof styles> {};

const Component = ({ classes }: Props) => {
  const page = useSelector(selectPage);
  const pokemons = useSelector(selectPokemonsItems);
  const pokemonsLoading = useSelector(selectPokemonsIsLoading);
  const caughtPokemons = useSelector(selectCaughtPokemonsItems);
  const caughtPokemonsLoading = useSelector(selectCaughtPokemonsIsLoading);
  const endOfPageRef = useRef(null);
  const dispatch = useDispatch();

  const handleNext = () => dispatch(fetchNextPagePokemons());

  const scrollPage = () => {
    const element = endOfPageRef.current as Element | null;
    if (element) {
      element.scrollIntoView({ behavior: "smooth" });
    }
  };

  const getPokemons = () => {
    dispatch(fetchPaginatedPokemons({ page }));
    if (!caughtPokemons.length) {
      dispatch(fetchCaughtPokemons());
    }
  };

  useEffect(() => {
    console.log("PokemonsPage-ComponentDidMount");
    if (!pokemons.length) {
      console.log("PokemonsPage-PokemonsEmpty-RETRIEVING!");
      getPokemons();
    }
  }, []);

  useEffect(() => {
    if (pokemonsLoading) return;
    scrollPage();
  }, [pokemonsLoading]);

  if (caughtPokemonsLoading) return <Loader showText />;

  console.log("PokemonsPage-Render");

  return (
    <div className={classes.root}>
      <Grid container spacing={24} justify="center">
        {
          pokemons.map(({ id, name }) => {
            const isAlreadyCaught = Boolean(caughtPokemons.find(({pokemonId}) => id === pokemonId));
            const handleCatch = () => dispatch(postAndAddCaughtPokemon(id, name));
            const cardActions = (
              <CardActions className={classes.actions}>
                <Button
                  disabled={isAlreadyCaught}
                  variant="outlined"
                  size="medium"
                  color="primary"
                  onClick={ isAlreadyCaught ? undefined : handleCatch }
                >
                  Catch
                </Button>
              </CardActions>
            );
            return (
              <PokemonItem key={id} pokemonId={id} name={name} cardActions={cardActions} link />
            );
          })
        }
      </Grid>
      <div className={classes.nextPageButtonWrapper} ref={endOfPageRef}>
        <Button
          className={classes.nextPageButton}
          disabled={pokemonsLoading}
          variant="contained"
          color="primary"
          onClick={handleNext}
        >Load more</Button>
        {
          pokemonsLoading && (<Loader className={classes.nextPageButtonLoader} size={24} />)
        }
      </div>
    </div>
  );
};

export const PokemonsPage = withStyles(styles)(Component);
