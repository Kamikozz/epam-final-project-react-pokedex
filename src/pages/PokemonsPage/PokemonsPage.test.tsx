import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { PokemonsPage } from './PokemonsPage';
import routes from '../../routes';
import store from "../../store";

beforeAll(() => {
  window.HTMLElement.prototype.scrollIntoView = jest.fn();
});

it('renders without crashing', () => {
  render(
    <Provider store={store}>
      <Router>
        <Switch>
          <Route path={routes.pokemonsPage} component={PokemonsPage} />
        </Switch>
      </Router>
    </Provider>
  );
});

it('matches snapshot', () => {
  const { container } = render(
    <Provider store={store}>
      <Router>
        <Switch>
          <Route path={routes.pokemonsPage} component={PokemonsPage} />
        </Switch>
      </Router>
    </Provider>
  );
  expect(container.firstChild).toMatchSnapshot()
});

