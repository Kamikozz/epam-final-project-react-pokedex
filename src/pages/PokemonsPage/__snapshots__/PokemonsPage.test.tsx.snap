// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="Loader-progressColor-150 "
>
  <div>
    LOADING
  </div>
  <div
    class="MuiCircularProgress-root-151 MuiCircularProgress-indeterminate-153 Loader-progress-149"
    role="progressbar"
    style="width: 70px; height: 70px;"
  >
    <svg
      class="MuiCircularProgress-svg-156"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle-157 MuiCircularProgress-circleIndeterminate-159"
        cx="44"
        cy="44"
        fill="none"
        r="21.5"
        stroke-width="1"
      />
    </svg>
  </div>
</div>
`;
